make rom_report fails for qemu_x86 (not finding zephyr.bin)

Previously, rom_report and ram_report targets worked for the mqtt_publisher example.
Now, it seems these targets are broken (or, is there a new usage?)


```
zephyr[master]> cd samples/net/mqtt_publisher/
mqtt_publisher[master]> mdkir build; cd build
build[master]> cmake .. -DBOARD=qemu_x86
build[master]> make rom_report
[...]
[100%] Building C object zephyr/CMakeFiles/kernel_elf.dir/misc/empty_file.c.obj
[100%] Linking C executable zephyr.elf
Generating files from zephyr.elf for board: qemu_x86
[100%] Built target kernel_elf
Scanning dependencies of target rom_report
Traceback (most recent call last):
  File "/home/gpitney/zephyr/scripts/footprint/size_report", line 356, in <module>
    0, 0)
  File "/home/gpitney/zephyr/scripts/footprint/size_report", line 94, in get_footprint_from_bin_and_statfile
    total_used_flash = os.path.getsize(bin_file)
  File "/usr/lib/python3.4/genericpath.py", line 50, in getsize
    return os.stat(filename).st_size
FileNotFoundError: [Errno 2] No such file or directory: '/home/gpitney/zephyr/samples/net/mqtt_publisher/build/zephyr/zephyr.bin'
zephyr/cmake/reports/CMakeFiles/rom_report.dir/build.make:57: recipe for target 'zephyr/cmake/reports/CMakeFiles/rom_report' failed
make[3]: *** [zephyr/cmake/reports/CMakeFiles/rom_report] Error 1
CMakeFiles/Makefile2:2582: recipe for target 'zephyr/cmake/reports/CMakeFiles/rom_report.dir/all' failed
make[2]: *** [zephyr/cmake/reports/CMakeFiles/rom_report.dir/all] Error 2
CMakeFiles/Makefile2:2589: recipe for target 'zephyr/cmake/reports/CMakeFiles/rom_report.dir/rule' failed
make[1]: *** [zephyr/cmake/reports/CMakeFiles/rom_report.dir/rule] Error 2
Makefile:664: recipe for target 'rom_report' failed
make: *** [rom_report] Error 2

build[master]> find . -name 'zephyr.bin'
build[master]> 
```

--- a/cmake/reports/CMakeLists.txt
+++ b/cmake/reports/CMakeLists.txt
@@ -8,6 +8,7 @@ foreach(report ram_report rom_report)
     ${ZEPHYR_BASE}/scripts/footprint/size_report
     ${flag_for_${report}}
     --objdump ${CMAKE_OBJDUMP}
+    --objcopy ${CMAKE_OBJCOPY}
     --nm      ${CMAKE_NM}
     -o ${PROJECT_BINARY_DIR}
     DEPENDS ${logical_target_for_zephyr_elf}

--- a/scripts/footprint/size_report
+++ b/scripts/footprint/size_report
@@ -31,6 +31,8 @@ parser.add_option("-F", "--rom",
                   help="print ROM statistics")
 parser.add_option("-s", "--objdump", type="string", dest="bin_objdump",
                   help="Path to the GNU binary utility objdump")
+parser.add_option("-c", "--objcopy", type="string", dest="bin_objcopy",
+                  help="Path to the GNU binary utility objcopy")
 parser.add_option("-n", "--nm", type="string", dest="bin_nm",
                   help="Path to the GNU binary utility nm")
 
@@ -406,14 +408,18 @@ def print_tree(data, total, depth):
     print('{:92d}'.format(total))
     return totp
 
+bin_file = os.path.join(options.outdir, options.binary + ".bin")
+stat_file = os.path.join(options.outdir, options.binary + ".stat")
+elf_file = os.path.join(options.outdir, options.binary + ".elf")
 
-binary = os.path.join(options.outdir, options.binary + ".elf")
+if not os.path.exists(bin_file):
+    FNULL = open(os.devnull, 'w')
+    subprocess.call([options.bin_objcopy,"-S", "-Obinary", "-R", ".comment", "-R",
+                   "COMMON", "-R", ".eh_frame", elf_file, bin_file],
+                   stdout=FNULL, stderr=subprocess.STDOUT)
 
-if options.outdir and os.path.exists(binary):
-    fp = get_footprint_from_bin_and_statfile(
-        "%s/%s.bin" % (options.outdir, options.binary),
-        "%s/%s.stat" % (options.outdir, options.binary),
-        0, 0)
+if options.outdir and os.path.exists(elf_file):
+    fp = get_footprint_from_bin_and_statfile(bin_file, stat_file, 0, 0)
     base = os.environ['ZEPHYR_BASE']
     ram, data = generate_target_memory_section(
         options.bin_objdump, options.bin_nm, options.outdir, options.binary,
@@ -424,4 +430,4 @@ if options.outdir and os.path.exists(binary):
         print_tree(ram, fp['total_ram'], options.depth)
 
 else:
-    print("%s does not exist." % (binary))
+    print("%s does not exist." % (elf_file))

--- a/scripts/footprint/size_report
+++ b/scripts/footprint/size_report
@@ -16,27 +16,6 @@ import operator
 import platform
 from pathlib import Path
 
-parser = OptionParser()
-parser.add_option("-d", "--depth", dest="depth", type="int",
-                  help="How deep should we go into the tree", metavar="DEPTH")
-parser.add_option("-o", "--outdir", dest="outdir",
-                  help="read files from directory OUT", metavar="OUT")
-parser.add_option("-k", "--kernel-name", dest="binary", default="zephyr",
-                  help="kernel binary name")
-parser.add_option("-r", "--ram",
-                  action="store_true", dest="ram", default=False,
-                  help="print RAM statistics")
-parser.add_option("-F", "--rom",
-                  action="store_true", dest="rom", default=False,
-                  help="print ROM statistics")
-parser.add_option("-s", "--objdump", type="string", dest="bin_objdump",
-                  help="Path to the GNU binary utility objdump")
-parser.add_option("-c", "--objcopy", type="string", dest="bin_objcopy",
-                  help="Path to the GNU binary utility objcopy")
-parser.add_option("-n", "--nm", type="string", dest="bin_nm",
-                  help="Path to the GNU binary utility nm")
-
-(options, args) = parser.parse_args()
 
 # Return a dict containing {
 #   symbol_name: {:,path/to/file}/symbol
@@ -408,26 +387,54 @@ def print_tree(data, total, depth):
     print('{:92d}'.format(total))
     return totp
 
-bin_file = os.path.join(options.outdir, options.binary + ".bin")
-stat_file = os.path.join(options.outdir, options.binary + ".stat")
-elf_file = os.path.join(options.outdir, options.binary + ".elf")
 
-if not os.path.exists(bin_file):
-    FNULL = open(os.devnull, 'w')
-    subprocess.call([options.bin_objcopy,"-S", "-Obinary", "-R", ".comment", "-R",
-                   "COMMON", "-R", ".eh_frame", elf_file, bin_file],
-                   stdout=FNULL, stderr=subprocess.STDOUT)
+def main():
+    parser = OptionParser()
+    parser.add_option("-d", "--depth", dest="depth", type="int",
+                      help="How deep should we go into the tree", metavar="DEPTH")
+    parser.add_option("-o", "--outdir", dest="outdir",
+                      help="read files from directory OUT", metavar="OUT")
+    parser.add_option("-k", "--kernel-name", dest="binary", default="zephyr",
+                      help="kernel binary name")
+    parser.add_option("-r", "--ram",
+                      action="store_true", dest="ram", default=False,
+                      help="print RAM statistics")
+    parser.add_option("-F", "--rom",
+                      action="store_true", dest="rom", default=False,
+                      help="print ROM statistics")
+    parser.add_option("-s", "--objdump", type="string", dest="bin_objdump",
+                      help="Path to the GNU binary utility objdump")
+    parser.add_option("-c", "--objcopy", type="string", dest="bin_objcopy",
+                      help="Path to the GNU binary utility objcopy")
+    parser.add_option("-n", "--nm", type="string", dest="bin_nm",
+                      help="Path to the GNU binary utility nm")
 
-if options.outdir and os.path.exists(elf_file):
-    fp = get_footprint_from_bin_and_statfile(bin_file, stat_file, 0, 0)
-    base = os.environ['ZEPHYR_BASE']
-    ram, data = generate_target_memory_section(
-        options.bin_objdump, options.bin_nm, options.outdir, options.binary,
-        base + '/', None)
-    if options.rom:
-        print_tree(data, fp['total_flash'], options.depth)
-    if options.ram:
-        print_tree(ram, fp['total_ram'], options.depth)
+    (options, args) = parser.parse_args()
 
-else:
-    print("%s does not exist." % (elf_file))
+    bin_file = os.path.join(options.outdir, options.binary + ".bin")
+    stat_file = os.path.join(options.outdir, options.binary + ".stat")
+    elf_file = os.path.join(options.outdir, options.binary + ".elf")
+
+    if not os.path.exists(bin_file):
+        FNULL = open(os.devnull, 'w')
+        subprocess.call([options.bin_objcopy,"-S", "-Obinary", "-R", ".comment", "-R",
+                       "COMMON", "-R", ".eh_frame", elf_file, bin_file],
+                       stdout=FNULL, stderr=subprocess.STDOUT)
+
+    if options.outdir and os.path.exists(elf_file):
+        fp = get_footprint_from_bin_and_statfile(bin_file, stat_file, 0, 0)
+        base = os.environ['ZEPHYR_BASE']
+        ram, data = generate_target_memory_section(
+            options.bin_objdump, options.bin_nm, options.outdir, options.binary,
+            base + '/', None)
+        if options.rom:
+            print_tree(data, fp['total_flash'], options.depth)
+        if options.ram:
+            print_tree(ram, fp['total_ram'], options.depth)
+
+    else:
+        print("%s does not exist." % (elf_file))
+
+
+if __name__ == "__main__":
+    main()

